File: my_tests/local_script_test1.py
import json
import yaml
from pathlib import Path
import numpy as np
import utility

def load_json_data(directory: Path, json_file: str):
    '''Load and parse a JSON file from the given directory, returns a plain dict.'''
    path = directory / json_file
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)
    
def load_yaml_data(dir, yaml_file):
    '''Load a YAML file from the given directory, returns a plain dict.'''
    path = dir / yaml_file
    with open(path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)
    

def main():
    # Resolve this script's directory (so paths work no matter where the file is run from)
    script_dir = Path(__file__).resolve().parent

    # Load the environment data
    env_dict = load_json_data(script_dir, "env.json")
    env = utility.objectview(env_dict)
    map_data = utility.objectview(env.map)
    dims = map_data.dimensions
    agents = env.agents
    n_agents = len(agents)

    # Load the schedule data
    schedule_dict = load_json_data(script_dir, "schedule.json")
    agent_schedule = schedule_dict["schedule"]
    agent_names = list(agent_schedule.keys())

    # Load the background image
    img = utility.load_background(script_dir, "warehouse_image.png")

    # Load the experiment parameters
    config_dict = load_yaml_data(script_dir, "config.yaml")
    config = utility.objectview(config_dict)

    N = int(config.N)                      # Number of robots utilized for the experiment
    exp_time = float(config.exp_time)      # Total experiment time in seconds
    limits = np.array([config.limits["x_lim"], config.limits["y_lim"]], dtype=float)     # Robotarium arena half-lengths in meters
    img_e = np.array(config.img_e, dtype=float)       # [x_offset, y_offset]
    waypoint_e = np.array(config.waypoint_e, dtype=float)

    control_params = utility.objectview(config.control_params)
    vx_K = control_params.si_x_vel_gain       # XVelocityGain(SI)
    vy_K = control_params.si_y_vel_gain       # YVelocityGain(SI)
    v_K = control_params.uni_lin_vel_gain     # LinearVelocityGain(Unicycle)
    w_lim = control_params.uni_ang_vel_limit  # AngularVelocityLimit(Unicycle)
    safe_r = control_params.safe_rad          # Safe radius
    control_params = np.array([vx_K, vy_K, v_K, w_lim, safe_r], dtype=float)

    # Build waypoints from Neural ATTF schedule data
    if not agent_names:
        raise ValueError("No agents found in schedule.json")

    first_agent_waypoints = agent_schedule[agent_names[0]]
    L = len(first_agent_waypoints)
    waypoints_A = np.zeros((L, 2, n_agents), dtype=float)
    for i in range(L):
        for j, name in enumerate(agent_names):
            wp_dict = agent_schedule[name][i]
            waypoints_A[i, 0, j] = float(wp_dict["x"])
            waypoints_A[i, 1, j] = float(wp_dict["y"])

    # Transform waypoints from the warehouse frame to the Robotarium frame
    waypoints_R = utility.transform_A_to_R(waypoints_A, dims, limits, waypoint_e)

    # Interpolate trajectory points at 30 Hz (each Robotarium time step)
    traj = utility.linear_interpolator(waypoints_R, freq=30)

    # Set the background and perform the experiment
    utility.perform_experiment(N, control_params, traj, img, exp_time, limits, img_e)


if __name__ == "__main__":
    main()

File: my_tests/local_script_test1.m
% Robotarium Experiment

%% Resolve this script's directory (so paths work no matter where you run from)
script_dir = fileparts(mfilename('fullpath'));

%% Load the environment (env.json)
env = load_json_data(script_dir, 'env.json');
agents = env.agents;
n_agents = numel(agents);
mapData = env.map;
dims = mapData.dimensions;

% Other information can also be retrieved from "env" as below.
% delays = env.delays;
% scale = mapData.scale;
% goal_locations = mapData.goal_locations;
% non_task_endpoints = mapData.non_task_endpoints;
% obstacles = mapData.obstacles;
% start_locations = mapData.start_locations;
% n_delays_per_agent = env.n_delays_per_agent;
% n_tasks = env.n_tasks;
% task_freq = env.task_freq;
% tasks = env.tasks;


%% Load the schedule data (schedule.json)
schedule = load_json_data(script_dir, 'schedule.json');
agent_schedule = schedule.schedule;
agent_names = fieldnames(agent_schedule);

% Other information can also be retrieved from "schedule" as below.
% completed_tasks_times = schedule.completed_tasks_times;
% cost = schedule.cost;
% dyn_ob_path = schedule.dyn_ob_path;
% n_replans = schedule.n_replans;
% schedule_agent1 = agent_schedule.agent1;
% schedule_agent2 = agent_schedule.agent2;
% schedule_agent3 = agent_schedule.agent3;
% schedule_agent4 = agent_schedule.agent4;
% schedule_agent5 = agent_schedule.agent5;
% schedule_agent6 = agent_schedule.agent6;


%% Load the warehouse image
img = load_background(script_dir, 'warehouse_image.png');


%% Load the experiment parameters
config = load_json_data(script_dir, 'config.json');
N = config.N; % Number of robots utilized for the experiment
exp_time = config.exp_time; % Total experiment time in seconds
limits = config.limits; % Robotarium boundary limit in meters
img_e = config.img_e; % Background shift from the center in the Robotarium frame
waypoint_e = config.waypoint_e; % Waypoints shift from the center in Robotarium

% control_params(1): XVelocityGain in SI dynamics
% control_params(2): YVelocityGain in SI dynamics
% control_params(3): LinearVelocityGain in Unicycle dynamics
% control_params(4): AngularVelocityLimit in Unicycle dynamics
% control_params(5): Safe radius
control_params = config.control_params;


%% Trajectory calculation for all the robots in the Robotarium frame
% Loading waypoints from the loaded Neural ATTF data
waypoints_A = zeros(2, n_agents, size(agent_schedule.agent1,1));
for i = 1:size(waypoints_A, 3)
    for j = 1:size(waypoints_A, 2)
        waypoints_A(1, j, i) = agent_schedule.(agent_names{j})(i).x;
        waypoints_A(2, j, i) = agent_schedule.(agent_names{j})(i).y;
    end
end

% Waypoints tranfromation from Neural ATTF frame to Robotarium frame
waypoints_R = utility.transform_A_to_R(waypoints_A, dims, limits, waypoint_e);

% Interpolating trajectory points at 30Hz (each Robotarium time step)
traj = utility.linear_interpolator(waypoints_R);


%% Set the background and perform the experiment
utility.perform_experiment(N, control_params, exp_time, traj, img, limits, img_e);


%% Function Definitions
% -------------------------------------------------------------------------
% Function to load data from .json files
function data = load_json_data(dir, json_file)
    path = fullfile(dir, json_file);
    text = fileread(path);
    data = jsondecode(text);
end

% -------------------------------------------------------------------------
% Function to load background image
function img = load_background(dir, img_file)
    path = fullfile(dir, img_file);
    if exist(path, 'file')
        img = imread(path);
    else
        warning('Image file not found. Skipping background.');
        img = [];
    end
end